### PR TITLE
b/182218056: support downstream mTLS

### PR DESCRIPTION
**Pull request by TAOXUY**
**context:** current we already have supported downstream TLS, upstream mTLS and sidestream TLS.

**content:**
- support downstream client cert-based auth to make downstream mTLS
- add unit tests and integration test
- fix our cert generating make target

### CODE REVIEW

File: docker/generic/start_proxy.py [hunk 1]
```diff
@@ -182,6 +182,13 @@ def make_argparser():
         Cipher suites to use for downstream connections as a comma-separated list.
         Please refer to https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/common.proto#auth-tlsparameters''')
 
+    parser.add_argument('--ssl_server_root_cert_path', default=None, help='''
+         The file path of root certificates that ESPv2 uses to do downstream client certificate-based authentication.
+        If not specified, ESPv2 doesn't it by default.
```
Comment (qiwzhang) [inline docker/generic/start_proxy.py:187]: s/ESPv2 doesn't it by default/ESPv2 doens't verify client certificates/?

File: docker/generic/start_proxy.py [hunk 2]
```diff
@@ -182,6 +182,13 @@ def make_argparser():
         Cipher suites to use for downstream connections as a comma-separated list.
         Please refer to https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/auth/common.proto#auth-tlsparameters''')
 
+    parser.add_argument('--ssl_server_root_cert_path', default=None, help='''
```
Comment (qiwzhang) [inline docker/generic/start_proxy.py:185]: mTLS means:
* verify client certificates
* specify which clients are allowed.  Do we want to support that? or it is in a separate pr?

File: src/go/configmanager/flags/flags.go [hunk 3]
```diff
@@ -60,6 +60,7 @@ var (
 
 	SslServerCertPath                = flag.String("ssl_server_cert_path", "", "Path to the certificate and key that ESPv2 uses to act as a HTTPS server")
 	SslServerCipherSuites            = flag.String("ssl_server_cipher_suites", "", "Cipher suites to use for downstream connections as a comma-separated list.")
+	SslServerRootCertsPath           = flag.String("ssl_server_root_cert_path", "", "The file path of root certificates that ESPv2 uses to verify downstream client certificate. If not specified, ESPv2 doesn't the validation by default.")
```
Comment (qiwzhang) [inline src/go/configmanager/flags/flags.go:63]: ESPv2 doens't verify client certificate.

File: docker/generic/start_proxy.py (hunk 1 above)
Comment (timothyleung1) [reply docker/generic/start_proxy.py:187]: Not sure if this is better: The file path of root certificates that ESPv2 uses to verify downstream client certificate.

File: src/go/util/transport_socket_test.go [hunk 4]
```diff
@@ -188,6 +189,42 @@ func TestCreateDownstreamTransportSocket(t *testing.T) {
...
 			} `,
 		},
+		{
+			desc:               "Downstream Transport Socket for mTLS",
+			sslPath:            "/etc/ssl/endpoints/",
+			sslRootCertPath:    "/etc/ssl/endpoints/root.crt",
+			sslMinimumProtocol: "TLSv1.1",
+			wantTransportSocket: `{
+				"name": "envoy.transport_sockets.tls",
+				"typedConfig": {
```
Comment (timothyleung1) [inline src/go/util/transport_socket_test.go:199]: Im not 100% familar with ESPv2 yet, but should there be a `require_client_certificate` flag in the config? https://www.envoyproxy.io/docs/envoy/latest/start/quick-start/securing

**Review by timothyleung1 — COMMENTED** [review]
Thanks for working on this!

File: src/go/util/transport_socket_test.go (hunk 4 above)
Comment (timothyleung1) [reply src/go/util/transport_socket_test.go:199]: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#envoy-v3-api-msg-extensions-transport-sockets-tls-v3-certificatevalidationcontext And are we going to support all kind of client auth or just trustedCA?

File: docker/generic/start_proxy.py (hunk 1 above)
Comment (TAOXUY) [reply docker/generic/start_proxy.py:187]: Done

File: src/go/configmanager/flags/flags.go (hunk 3 above)
Comment (TAOXUY) [reply src/go/configmanager/flags/flags.go:63]: Done.

File: docker/generic/start_proxy.py [hunk 5]
```diff
@@ -182,6 +182,13 @@ def make_argparser():
...
+         The file path of root certificates that ESPv2 uses to do downstream client certificate-based authentication.
+        If not specified, ESPv2 doesn't it by default. 
+        
+        Before using this feature, please verify if your compute environment is compatible with it.
```
Comment (qiwzhang) [inline docker/generic/start_proxy.py:189]: what computer environment?   how to verify?, verify what?

File: src/go/configmanager/flags/flags.go (hunk 3 above)
Comment (qiwzhang) [reply src/go/configmanager/flags/flags.go:63]: I still see this "ESPv2 doesn't the validation by default".  This is not right grammar

File: docker/generic/start_proxy.py (hunk 2 above)
Comment (TAOXUY) [reply docker/generic/start_proxy.py:185]: You are right. This PR, like our existing TLS connections, only do the basic server certs validation/authn. As for tls access control, envoy does have a bunch of other [options](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext) to do access control or extra validation. I am thinking we make another ones if necessary after asking them what they need. We can even encourage to contribute based on this PR.

File: src/go/util/transport_socket_test.go (hunk 4 above)
Comment (TAOXUY) [reply src/go/util/transport_socket_test.go:199]: `require_client_certifacte` added. It is interesting that seems this field doesn't work and  client certificate  is enforced. As I didn't set it before, my test would raise error if the client doesn't provide certs, which fits https://github.com/envoyproxy/envoy/issues/17224.

Yes, this PR only provides basic client cert authn for this PR and for those [options](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-certificatevalidationcontext), what do you need?

File: tests/integration_test/transport_security_test/transport_security_test.go [hunk 6]
```diff
@@ -162,73 +163,182 @@ func TestServiceControlWithTLS(t *testing.T) {
...
 		},
 		{
-			desc:         "Succcess for HTTP2 client with TLS",
-			httpsVersion: 2,
+			desc:                  "failure(client authn),  client cert is unknown to proxy's root cert",
+			scheme:                "http",
+			httpVersions:          []int{1},
+			sslServerRootCertPath: platform.GetFilePath(platform.DownstreamClientCert),
+			getClientCred: func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				c, err := tls.LoadX509KeyPair(platform.GetFilePath(platform.MismatchCert), platform.GetFilePath(platform.MismatchKey))
```
Comment (qiwzhang) [inline tests/integration_test/transport_security_test/transport_security_test.go:198]: It seems the code for getClientCred are all the same,  just different pairs of client_cert files.  Maybe moving this code into client.DoHttpsGet

Comment (TAOXUY) [reply tests/integration_test/transport_security_test/transport_security_test.go]: Done.

File: docker/generic/start_proxy.py (hunk 5 above)
Comment (TAOXUY) [reply docker/generic/start_proxy.py]: Just reworded. I wanna warn users that many GCP platforms will terminate TSL/mTSL in ingress.

File: src/go/util/transport_socket_test.go (hunk 4 above)
Comment (timothyleung1) [reply src/go/util/transport_socket_test.go:199]: im having a meeting with partner team on Wednesday and ill raise that point. but i guess this CL is a good start already, it should be easy to add new options in the future.

**Review by qiwzhang — APPROVED** [review]

File: src/go/util/transport_socket_test.go (hunk 4 above)
Comment (TAOXUY) [reply src/go/util/transport_socket_test.go:199]: SG

File: docker/generic/start_proxy.py [hunk 7]
```diff
@@ -176,12 +176,28 @@ def make_argparser():
     parser.add_argument('--ssl_server_cert_path', default=None, help='''
         Proxy's server cert path. When configured, ESPv2 only accepts HTTP/1.x and
         HTTP/2 secure connections on listener_port. Requires the certificate and
-        key files "server.crt" and "server.key" within this path.''')
+        key files "server.crt" and "server.key" within this path.
+        
+        Before using this feature, please make sure TLS isn't terminated before ESPv2
+        in your deployment model. In general, Cloud Run, GKE(GCLB enforced in ingress)
```
Comment (nareddyt) [inline docker/generic/start_proxy.py:182]: GKE with container-native load balancing

File: tests/env/platform/ports.go [hunk 8]
```diff
@@ -32,12 +32,12 @@ const (
 	TestAddHeaders
```
Comment (nareddyt) [inline tests/env/platform/ports.go:32]: Can you add a comment on how you are sorting these automatically?

File: tests/env/testdata/mismatch.crt [hunk 9]
```diff
@@ -0,0 +1,23 @@
+-----BEGIN CERTIFICATE-----
```
Comment (nareddyt) [inline tests/env/testdata/mismatch.crt:1]: Need some information on how these mismatch certs were generated. I cannot tell what the "mismatch" is looking at this file

File: tests/integration_test/transport_security_test/transport_security_test.go [hunk 10]
```diff
@@ -162,73 +162,152 @@ func TestServiceControlWithTLS(t *testing.T) {
...
-			t.Errorf("Test (%s): failed\nexpected: %v\ngot: %v", tc.desc, tc.wantError, err)
-		}
+
+			s := env.NewTestEnv(platform.TestDownstreamMTLS, platform.EchoSidecar)
+			defer s.TearDown(t)
+			if err := s.Setup(args); err != nil {
+				t.Fatalf("fail to setup test env, %v", err)
+			}
+
+			// FIXME: Use of localhost. Difficult to generate certs with ip addresses.
```
Comment (nareddyt) [inline tests/integration_test/transport_security_test/transport_security_test.go:288]: Yeah this will not work in google3 easily. That's fine for now.

**Review by nareddyt — APPROVED** [review]

File: docker/generic/start_proxy.py (hunk 7 above)
Comment (TAOXUY) [reply docker/generic/start_proxy.py:182]: Done.

File: tests/env/platform/ports.go (hunk 8 above)
Comment (TAOXUY) [reply tests/env/platform/ports.go:32]: I wish I could but it is done by IDE.

File: tests/env/testdata/mismatch.crt (hunk 9 above)
Comment (TAOXUY) [reply tests/env/testdata/mismatch.crt:1]: Mismatched crt is used in the case when the server crt is unknown to the client root ca. Before, we used proxy.crt to replace server.crt or vice versa to make this case but it is confusing.

**Pull request merged**